File: app/src/renderer/src/api/http.ts
/* The transport, and nothing about any screen. Every URL is relative: in dev Vite
   proxies /admin and /api to the daemon, in prod the desktop shell does. Domain calls
   belong next to the view that makes them. */

export class HttpError extends Error {
  constructor(
    readonly status: number,
    readonly detail: string
  ) {
    super(`${status}: ${detail}`)
    this.name = 'HttpError'
  }
}

/* The house refusal is FastAPI's `{"detail": ...}`; the dialect routes answer
   `{"error": {"message": ...}}`. Both end up in `HttpError.detail`. */
async function refuse(answer: Response): Promise<never> {
  let detail = answer.statusText
  try {
    const body: unknown = await answer.json()
    if (body && typeof body === 'object') {
      const shape = body as { detail?: unknown; error?: { message?: unknown } }
      if (typeof shape.detail === 'string') detail = shape.detail
      else if (typeof shape.error?.message === 'string') detail = shape.error.message
      else detail = JSON.stringify(body)
    }
  } catch {
    /* a body that is not JSON leaves the status text standing */
  }
  throw new HttpError(answer.status, detail)
}

export async function json<T>(path: string, init?: RequestInit): Promise<T> {
  const answer = await fetch(path, {
    ...init,
    headers:
      init?.body === undefined
        ? init?.headers
        : { 'content-type': 'application/json', ...init?.headers }
  })
  if (!answer.ok) await refuse(answer)
  if (answer.status === 204) return undefined as T
  return (await answer.json()) as T
}

export function send<T>(method: string, path: string, body?: unknown): Promise<T> {
  return json<T>(path, {
    method,
    ...(body === undefined ? {} : { body: JSON.stringify(body) })
  })
}

/* Every stream this daemon serves is `text/event-stream` carrying whole JSON states
   on `data:` lines, plus `: keep-alive` comments that mean nothing but the connection
   is warm. `fetch` rather than `EventSource` because half of them are POSTs and all of
   them need an abort that reaches the socket. */
export async function* sse(path: string, init?: RequestInit): AsyncGenerator<string> {
  const answer = await fetch(path, {
    ...init,
    headers:
      init?.body === undefined
        ? init?.headers
        : { 'content-type': 'application/json', ...init?.headers }
  })
  if (!answer.ok) await refuse(answer)
  if (answer.body === null) return
  const lines = answer.body.pipeThrough(new TextDecoderStream()).getReader()
  let held = ''
  try {
    for (;;) {
      const { value, done } = await lines.read()
      if (done) break
      held += value
      /* A frame ends at the blank line; a chunk may carry half of one. */
      let cut = held.indexOf('\n\n')
      while (cut !== -1) {
        const frame = held.slice(0, cut)
        held = held.slice(cut + 2)
        const data = frame
          .split('\n')
          .filter((line) => line.startsWith('data:'))
          .map((line) => line.slice(5).trimStart())
          .join('\n')
        if (data) yield data
        cut = held.indexOf('\n\n')
      }
    }
  } finally {
    await lines.cancel().catch(() => {})
  }
}

export async function* events<T>(path: string, init?: RequestInit): AsyncGenerator<T> {
  for await (const data of sse(path, init)) yield JSON.parse(data) as T
}

/* ── jobs ─────────────────────────────────────────────────────────────── */

export type JobState = 'pending' | 'running' | 'ok' | 'error' | 'cancelled'

export interface Progress {
  message: string
  completed: number
  /* Absent while unknown: a download learns its size from the first response. */
  total: number | null
}

interface JobBase {
  id: string
  state: JobState
  progress: Progress
  created_at: number
  updated_at: number
  error: string | null
}

/* `kind` is the tag `subject` is read back with, on this side too: a screen that draws a job
   beside the thing it acts on gets there by narrowing, not by parsing `progress.message`. */
export type JobView =
  | (JobBase & { kind: 'download'; subject: { model: string } })
  | (JobBase & { kind: 'load'; subject: { model: string } })
  | (JobBase & { kind: 'quantize'; subject: { model: string; target: string } })
  | (JobBase & { kind: 'bench'; subject: { model: string } })
  | (JobBase & { kind: 'benchmark'; subject: { kind: string; models: string[] } })

export const isFinished = (job: JobView): boolean =>
  job.state === 'ok' || job.state === 'error' || job.state === 'cancelled'

export const getJob = (id: string): Promise<JobView> => json<JobView>(`/admin/jobs/${id}`)

export const listJobs = (active = false): Promise<JobView[]> =>
  json<JobView[]>(`/admin/jobs${active ? '?active=true' : ''}`)

export const cancelJob = (id: string): Promise<JobView> => send<JobView>('DELETE', `/admin/jobs/${id}`)

/* Every frame is the whole state and the stream opens with the state as it is now, so
   a late subscriber loses nothing and a repeated frame costs nothing. It ends when the
   job reaches a terminal state. */
export const jobEvents = (id: string, signal?: AbortSignal): AsyncGenerator<JobView> =>
  events<JobView>(`/admin/jobs/${id}/events`, { signal })

File: packages/application/src/sideros_application/api/http.py
"""How this window reaches the daemon, and the two shapes an answer arrives in: a body,
or a stream of them.

The house refusal is FastAPI's `{"detail": ...}`, which is what a screen shows rather than
a status code.
"""

from __future__ import annotations

import json
import os
from collections.abc import AsyncGenerator
from typing import cast
from urllib.parse import quote

import httpx

BASE = os.environ.get("SIDEROS_DAEMON_URL", "http://127.0.0.1:8642").rstrip("/")


class Refused(Exception):
    pass


def client() -> httpx.AsyncClient:
    return httpx.AsyncClient(base_url=BASE, timeout=30)


def at(identifier: str) -> str:
    """A model id in a path segment. The route is a `:path`, so the slash in `org/name`
    stays a slash and only the rest is escaped."""
    return "/".join(quote(part, safe="") for part in identifier.split("/"))


async def get(path: str) -> object:
    async with client() as http:
        answer = await http.get(path)
        answer.raise_for_status()
        return answer.json()


async def send(method: str, path: str, body: object | None = None) -> object:
    async with client() as http:
        answer = await http.request(
            method, path, json=body, timeout=None if method == "POST" else 30
        )
        if answer.is_error:
            raise Refused(_detail(answer))
        return None if answer.status_code == 204 else answer.json()


async def events(http: httpx.AsyncClient, path: str) -> AsyncGenerator[dict[str, object]]:
    """text/event-stream carrying whole JSON states on `data:` lines, plus `: keep-alive`
    comments that mean nothing but the connection is warm."""
    async with http.stream("GET", path) as answer:
        answer.raise_for_status()
        held: list[str] = []
        async for line in answer.aiter_lines():
            if line.startswith("data:"):
                held.append(line[5:].lstrip())
            elif line == "" and held:
                yield cast(dict[str, object], json.loads("\n".join(held)))
                held = []


async def post_events(path: str, body: object) -> AsyncGenerator[str]:
    """The `data:` payloads of a stream a POST opens, verbatim — the dialect's own frames
    are not this module's to read, and `[DONE]` is one of them."""
    async with (
        httpx.AsyncClient(base_url=BASE, timeout=None) as http,
        http.stream("POST", path, json=body) as answer,
    ):
        if answer.is_error:
            await answer.aread()
            raise Refused(_detail(answer))
        held: list[str] = []
        async for line in answer.aiter_lines():
            if line.startswith("data:"):
                held.append(line[5:].lstrip())
            elif line == "" and held:
                yield "\n".join(held)
                held = []


def _detail(answer: httpx.Response) -> str:
    try:
        body = answer.json()
    except ValueError:
        return answer.reason_phrase
    if isinstance(body, dict):
        detail = body.get("detail")
        if isinstance(detail, str):
            return detail
        error = body.get("error")
        if isinstance(error, dict) and isinstance(error.get("message"), str):
            return str(error["message"])
    return answer.reason_phrase
